optimizely/config_manager.py

```diff
@@ -0,0 +1,233 @@
+            self.logger.debug('Invalid update_interval {} provided. Defaulting to {}'.format(
+                update_interval,
+                enums.ConfigManager.DEFAULT_UPDATE_INTERVAL)
+            )
+            self.update_interval = enums.ConfigManager.DEFAULT_UPDATE_INTERVAL
+
+    def set_last_modified(self, response_headers):
+        """ Looks up and sets last modified time based on Last-Modified header in the response.
+
+         Args:
+           response_headers: requests.Response.headers
+         """
+        self.last_modified = response_headers.get(enums.HTTPHeaders.LAST_MODIFIED)
+
+    def set_config(self, datafile):
+        """ Looks up and sets datafile and config based on response body.
+
+         Args:
+           datafile: JSON string representing the Optimizely project.
+         """
+        # TODO(ali): Add validation here to make sure that we do not update datafile and config if not a valid datafile.
+        self._datafile = datafile
+        # TODO(ali): Add notification listener.
+        self._config = project_config.ProjectConfig(self._datafile, self.logger, self.error_handler)
+        self.logger.info('Received new datafile and updated config.')
+
+    def get_config(self):
+        """ Returns instance of ProjectConfig.
+
+        Returns:
+          ProjectConfig.
+        """
+        return self._config
+
+    def _handle_response(self, response):
+        """ Helper method to handle response containing datafile.
+
+        Args:
+            response: requests.Response
+        """
+        try:
+            response.raise_for_status()
+        except requests_exceptions.HTTPError as err:
+            self.logger.error('Fetching datafile from {} failed. Error: {}'.format(self.datafile_url, str(err)))
+            return
+
+        # Leave datafile and config unchanged if it has not been modified.
+        if response.status_code == http_status_codes.not_modified:
+            self.logger.debug('Not updating config as datafile has not updated since {}.'.format(self.last_modified))
+            return
+
+        self.set_last_modified(response.headers)
+        self.set_config(response.text)
+
+    def fetch_datafile(self):
+        """ Fetch datafile and set ProjectConfig. """
+
+        request_headers = {}
+        if self.last_modified:
+            request_headers[enums.HTTPHeaders.IF_MODIFIED_SINCE] = self.last_modified
+
+        response = requests.get(self.datafile_url, headers=request_headers)
+        self._handle_response(response)
+
+    @property
+    def is_running(self):
+        """ Check if polling thread is alive or not. """
+        return self._polling_thread.is_alive()
+
+    def _run(self):
+        """ Triggered as part of the thread which fetches the datafile and sleeps until next update interval. """
+        while self.is_running:
+            self.fetch_datafile()
+            time.sleep(self.update_interval)
+
+    def start(self):
```

Comment:
We also need a way to stop the updates

Reply:
Because we making `polling_thread` a daemon thread, it stops when the main thread stops: https://docs.python.org/2/library/threading.html#threading.Thread.daemon and so we do not need a stop. It auto-stops. Also, I will be calling this start in the constructor itself when I integrate this with optimizely.Optimizely.